sql: error when scanning bytes[] with lib/pq

I am using v1.1.4

I am facing a scan error when trying to read a bytes[] with "corrupt data" (Note the `}` inside the first element):

```
cockroach sql --insecure -e "CREATE DATABASE test"
cockroach sql --insecure -e "create table t (id serial, data bytes[] null,  primary key (id));" --database=test
cockroach sql --insecure -e "insert into t values (1, ARRAY[b'1}', b'2']);" --database=test
go run main.go
panic: sql: Scan error on column index 1: pq: unable to parse array; unexpected '\'' at offset 5
```

Content of main.go

```
package main

import (
	"database/sql"
	"fmt"

	"github.com/lib/pq"
)

type Row struct {
	id   int64
	data pq.ByteaArray
}

func main() {
	connStr := fmt.Sprintf("postgresql://%s:%s@%s:%d/%s?sslmode=%s", "root", "root", "localhost", 26257, "test", "disable")
	conn, err := sql.Open("postgres", connStr)
	if err != nil {
		panic(err)
	}

	rows, err := conn.Query("SELECT * FROM t")
	if err != nil {
		panic(err)
	}

	for rows.Next() {
		var row Row
		if err := rows.Scan(&row.id, &row.data); err != nil {
			panic(err)
		} else {
			fmt.Println(row)
		}
	}
}
```

--- a/pkg/sql/logictest/testdata/logic_test/array
+++ b/pkg/sql/logictest/testdata/logic_test/array
@@ -35,6 +35,11 @@ SELECT ARRAY['a,', 'b{', 'c}', 'd', 'e f']
 ----
 {"a,","b{","c}","d","e f"}
 
+query T
+SELECT ARRAY['1}'::BYTES]
+----
+{"\\x317d"}
+
 # TODO(jordan): #16487
 # query T
 # SELECT ARRAY[e'g\x10h']
@@ -619,7 +624,7 @@ INSERT INTO a VALUES (ARRAY['foo','bar','baz'])
 query T
 SELECT b FROM a
 ----
-{'\x666f6f','\x626172','\x62617a'}
+{"\\x666f6f","\\x626172","\\x62617a"}
 
 statement ok
 DROP TABLE a

--- a/pkg/sql/sem/tree/datum.go
+++ b/pkg/sql/sem/tree/datum.go
@@ -1124,20 +1124,30 @@ func (d *DBytes) Max(_ *EvalContext) (Datum, bool) {
 // AmbiguousFormat implements the Datum interface.
 func (*DBytes) AmbiguousFormat() bool { return true }
 
-// Format implements the NodeFormatter interface.
-func (d *DBytes) Format(ctx *FmtCtx) {
-	f := ctx.flags
-	withQuotes := f.HasFlags(fmtWithinArray) || !f.HasFlags(FmtFlags(lex.EncBareStrings))
-	if withQuotes {
-		ctx.WriteByte('\'')
-	}
-	ctx.WriteString("\\x")
+func writeAsHexString(ctx *FmtCtx, d *DBytes) {
 	b := string(*d)
 	for i := 0; i < len(b); i++ {
 		ctx.Write(stringencoding.RawHexMap[b[i]])
 	}
-	if withQuotes {
-		ctx.WriteByte('\'')
+}
+
+// Format implements the NodeFormatter interface.
+func (d *DBytes) Format(ctx *FmtCtx) {
+	f := ctx.flags
+	if f.HasFlags(fmtWithinArray) {
+		ctx.WriteString(`"\\x`)
+		writeAsHexString(ctx, d)
+		ctx.WriteString(`"`)
+	} else {
+		withQuotes := !f.HasFlags(FmtFlags(lex.EncBareStrings))
+		if withQuotes {
+			ctx.WriteByte('\'')
+		}
+		ctx.WriteString("\\x")
+		writeAsHexString(ctx, d)
+		if withQuotes {
+			ctx.WriteByte('\'')
+		}
 	}
 }
 

